File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/transition/ExplorationTransition.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.transition;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.core.implementation.job.Context;
import org.openmole.core.implementation.mole.SubMoleExecution;
import org.openmole.core.implementation.task.ExplorationTask;
import org.openmole.core.implementation.tools.ClonningService;
import org.openmole.core.model.data.IVariable;
import org.openmole.core.model.plan.IExploredPlan;
import org.openmole.core.model.plan.IFactorValues;
import org.openmole.core.model.job.IContext;
import org.openmole.core.model.transition.IExplorationTransition;
import org.openmole.core.model.transition.IGenericTransition;
import org.openmole.core.model.capsule.IExplorationTaskCapsule;
import org.openmole.core.model.tools.IRegistryWithTicket;
import org.openmole.core.model.capsule.IGenericTaskCapsule;
import org.openmole.core.model.data.IData;
import org.openmole.core.model.data.IPrototype;
import org.openmole.core.model.job.ITicket;
import org.openmole.core.model.mole.IMoleExecution;
import org.openmole.core.model.mole.ISubMoleExecution;

public class ExplorationTransition extends GenericTransition<IExplorationTaskCapsule> implements IExplorationTransition {

    public ExplorationTransition(IExplorationTaskCapsule start, IGenericTaskCapsule<?, ?> end) {
        super(start, end);
    }

    @Override
    public synchronized void performImpl(IContext global, IContext context, ITicket ticket, Set<String> toClone, IMoleExecution moleExecution, ISubMoleExecution subMole) throws InternalProcessingError, UserBadDataError {

        IRegistryWithTicket<IGenericTransition, IContext> registry = moleExecution.getLocalCommunication().getTransitionRegistry();
        registry.register(this, ticket, context);

        IExploredPlan values = context.getValue(ExplorationTask.ExploredPlan.getPrototype());
        context.removeVariable(ExplorationTask.ExploredPlan.getPrototype().getName());

        ISubMoleExecution subSubMole = new SubMoleExecution(moleExecution, subMole);

        int size = 0;

        Iterator<IFactorValues> factorIt = values.iterator();

        while(factorIt.hasNext()) {
            IFactorValues value = factorIt.next();

            size++;
            subSubMole.incNbJobInProgress();

            ITicket newTicket = moleExecution.nextTicket(ticket);
            IContext destContext = new Context();

            Set<IData<?>> notFound = new HashSet<IData<?>>();

            for (IData<?> in : getEnd().getCapsule().getTask().getInput()) {
                //Check filtred here to avoid useless clonning
                if (context.contains(in.getPrototype()) && !getFiltred().contains(in.getPrototype().getName())) {

                    IVariable<?> varToClone = context.getVariable(in.getPrototype());
                    IVariable<?> var;

                    if(!in.getMode().isImmutable() && factorIt.hasNext()) {
                        var = ClonningService.clone(varToClone);
                    } else {
                        var = varToClone;
                    }

                    destContext.putVariable(var);
                } else {
                    notFound.add(in);
                }
            }


            for(IData data : notFound) {
                IPrototype prototype = data.getPrototype();
                Object val = value.getValue(data.getPrototype());

                if (val != null) {
                    if (prototype.getType().isAssignableFrom(val.getClass())) {
                        destContext.putVariable(prototype, val);
                    } else {
                        throw new UserBadDataError("Factor variable " + prototype.getName() + " of type " + val.getClass().getName() + " found but not compatible with expected type for input " + prototype.getType().getName());
                    }
                }
            }
            
            submitNextJobsIfReady(global, destContext, newTicket, toClone, moleExecution, subSubMole);
        }

        subSubMole.decNbJobInProgress(size);
    }

    @Override
    protected void plugStart() {
        getStart().plugOutputTransition(this);
    }
}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/transition/ExplorationTransition.scala
/*
 * Copyright (C) 2011 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.transition

import org.openmole.core.implementation.task.ExplorationTask._
import org.openmole.core.implementation.data._
import org.openmole.core.implementation.data.Context._
import org.openmole.core.implementation.mole.SubMoleExecution
import org.openmole.core.implementation.mole.Capsule._
import org.openmole.core.model.task.IExplorationTask
import org.openmole.core.model.task.ITask
import org.openmole.core.model.data.DataModeMask._
import org.openmole.core.model.transition.{ IExplorationTransition, ISlot, ICondition, IAggregationTransition }
import org.openmole.core.model.data.{ IPrototype, IContext, IVariable, IData }
import org.openmole.core.model.mole.{ ICapsule, ITicket, ISubMoleExecution }
import org.openmole.misc.eventdispatcher.EventDispatcher
import org.openmole.misc.exception.InternalProcessingError
import org.openmole.misc.exception.UserBadDataError
import org.openmole.misc.tools.obj.ClassUtils._
import org.openmole.misc.tools.service.Logger
import org.openmole.misc.tools.service.Priority
import scala.collection.mutable.HashSet
import scala.collection.mutable.ListBuffer

class ExplorationTransition(start: ICapsule, end: ISlot, condition: ICondition = ICondition.True, filtered: Iterable[String] = Iterable.empty[String]) extends Transition(start, end, condition, filtered) with IExplorationTransition {

  override def _perform(context: IContext, ticket: ITicket, subMole: ISubMoleExecution) = {
    val subSubMole = subMole.newChild

    registerAggregationTransitions(ticket, subSubMole)
    submitIn(context, ticket, subSubMole)
  }

  def submitIn(context: IContext, ticket: ITicket, subMole: ISubMoleExecution) = {
    val (factors, outputs) = start.outputs.partition(d ⇒ (d.mode is explore) && d.prototype.`type`.isArray)
    val typedFactors = factors.map(_.prototype.asInstanceOf[IPrototype[Array[Any]]])
    val values = typedFactors.toList.map(context.value(_).get.toIterable).transpose

    val endTask = end.capsule.taskOrException

    for (value ← values) {

      val newTicket = subMole.moleExecution.nextTicket(ticket)

      val variables = new ListBuffer[IVariable[_]]

      for (in ← outputs)
        context.variable(in.prototype) match {
          case Some(v) ⇒ variables += v
          case None ⇒
        }

      for ((p, v) ← typedFactors zip value) {
        val fp = p.fromArray
        if (fp.accepts(v)) variables += new Variable(fp, v)
        else throw new UserBadDataError("Found value of type " + v.asInstanceOf[AnyRef].getClass + " incompatible with prototype " + fp)
      }
      submitNextJobsIfReady(ListBuffer() ++ variables, newTicket, subMole)
    }

  }

  private def registerAggregationTransitions(ticket: ITicket, subMoleExecution: ISubMoleExecution) = {
    val alreadySeen = new HashSet[ICapsule]
    val toProcess = new ListBuffer[(ICapsule, Int)]
    toProcess += ((end.capsule, 0))

    while (!toProcess.isEmpty) {
      val cur = toProcess.remove(0)
      val capsule = cur._1
      val level = cur._2

      if (!alreadySeen(capsule)) {
        alreadySeen += capsule

        capsule.outputTransitions.foreach {
          case t: IAggregationTransition ⇒
            if (level > 0) toProcess += t.end.capsule -> (level - 1)
            else if (level == 0) {
              subMoleExecution.aggregationTransitionRegistry.register(t, ticket, new ListBuffer)
              EventDispatcher.listen(subMoleExecution, Priority.LOW, new AggregationTransitionAdapter(t), classOf[ISubMoleExecution.Finished])
            }
          case t: IExplorationTransition ⇒ toProcess += t.end.capsule -> (level + 1)
          case t ⇒ toProcess += t.end.capsule -> level
        }
      }
    }
  }

}
